### ISSUE TITLE
Doesn't match Rust expressions when using trailing commas

### ISSUE THREAD

**Issue by RReverser** [ast-grep/ast-grep#374]
Minimal repro:

```rust
foobar(
  1,
  2,
);
```

In the original, more complex example, it took some time to figure out why it wasn't matching against a pattern

```rust
foobar($X, $Y)
```

While eventually I found out that adding explicit comma helps.

I know that comma is parsed as a separate token, but AST-wise both should be semantically equivalent and ideally should match in patterns.

**Comment by HerringtonDarkholme** [comment]
Thanks for reporting!
This is actually a known issue because, under the hood, ast-grep is using tree-sitter's CST(concrete syntax tree) to match the pattern and code. I admit ast is a little bit misnomer.

I think I have earned enough knowledge about how to convert a CST to AST so the match against the expression will succeed. The strategy is to ignore trivia node (that is, nameless and not field-less node).

I will experiment more to see if using last will have unexpected fallout. If no bad impact is found, I will implement a true AST matching algorithm.

**Comment by RReverser** [comment]
I see, thanks!

> The strategy is to ignore trivia node (that is, nameless and not field-less node).

Wouldn't that accidentally skip meaningful commas as well? Or is that not a problem because those commas are mandatory and can be implied, as without them code would be invalid anyway?

**Comment by HerringtonDarkholme** [comment]
Let me clarify. Suppose we have a code like `f(a, b)` now. 

The CST and AST will look like the following table. See the playground [link](https://ast-grep.github.io/playground.html#eyJtb2RlIjoiUGF0Y2giLCJsYW5nIjoicnVzdCIsInF1ZXJ5IjoiY29uc29sZS5sb2coJE1BVENIKSIsImNvbmZpZyI6IiMgQ29uZmlndXJlIFJ1bGUgaW4gWUFNTFxucnVsZTpcbiAgYW55OlxuICAgIC0gcGF0dGVybjogaWYgKGZhbHNlKSB7ICQkJCB9XG4gICAgLSBwYXR0ZXJuOiBpZiAodHJ1ZSkgeyAkJCQgfVxuY29uc3RyYWludHM6XG4gICMgTUVUQV9WQVI6IHBhdHRlcm4iLCJzb3VyY2UiOiJmKDEsIDIpIn0=).

|CST| AST|
|---|---|
|<img width="267" alt="image" src="https://github.com/ast-grep/ast-grep/assets/2883231/daa3f397-6835-46a5-b823-abcdf979bfc1">|<img width="279" alt="image" src="https://github.com/ast-grep/ast-grep/assets/2883231/7b5b046d-4d6e-461c-a633-df002326c9a8">|

We are using CST now so the parentheses are also considered during matching. Though they do not affect the matching result anyway.

> Wouldn't that accidentally skip meaningful commas

If we match code by AST, the tree structure will already imply meaningful commas and we are safe to skip them.

However, the prerequisite is that the tree-sitter grammar should be well-formed so that any meaningful trivia nodes are **indeed** trivial. 

I'm not sure if all grammars obey such rules...

**Comment by HerringtonDarkholme** [comment]
AST is not sufficient though... For example the following JS code are decoded as the same AST....

AST:
<img width="1032" alt="image" src="https://github.com/ast-grep/ast-grep/assets/2883231/2831e226-59f0-4768-8ed7-1fac65038e31">

CST:
![image](https://github.com/ast-grep/ast-grep/assets/2883231/9ee2d883-43bc-4470-9409-4ebfeeb48880)

`get` keyword is only encoded in CST....

**Comment by RReverser** [comment]
> However, the prerequisite is that the tree-sitter grammar should be well-formed so that any meaningful trivia nodes are **indeed** trivial.

Ah yeah that's the bit I was trying to clarify. Makes sense to assume it's already well-formed - that does make matching easier.

**Comment by HerringtonDarkholme** [comment]
I have come up with a smarter algorithm to match pattern and code. However, I think this will make ast-grep's pattern matching harder to learn.

## named nodes vs unnamed nodes
The story starts with the difference between **named node** and **unnamed node**. 

In tree-sitter, a node with `kind` is a named node. A node without `kind` is unnamed node. Conceptually, tree-sitter's output of all named nodes is equivalent to the source code's AST. If we add back unnamed node to tree-sitter's output, it becomes the full CST of the code.

## ast-grep's algorithm now
Currently, ast-grep matches the full CST: we matches all named nodes and unnamed nodes in the pattern with those in the source code.

However, this makes bad cases like trailing commas mentioned above. 
Maybe we can make ast-grep smarter by only looking at named node?  Unfortunately, this also does not work because most tree-sitter's parsers will not encode all semantics in named nodes. For example, these two pieces of code has exactly the same named node structure (a.k.a. AST.)
```
class A { method() {} }
class A { get method() {} }
```

The critical token `get` is parsed as an unnamed node. It is not trivial though! 

We can alleviate this problem by counting more nodes as **significant**. A possible way is that a node is significant if a node is named node or is an unnamed node with `field` relative to its parent.

Example:

```
a + b 
```
will be encoded as 
<img width="327" alt="image" src="https://github.com/ast-grep/ast-grep/assets/2883231/bdb1a220-2264-4b15-8893-362bf9bc0981">

Note `a` and `b` are named nodes, `identifier`. `+` is an unnamed node, but it is at the `op` field of `binary_expression`.

## improved algorithm by skipping insignificant nodes in source

Ok, we have a lot of background knowledge here. Now we can state our improved algorithm

We can match all nodes in pattern, but if the source has redundant insignificant node (unnamed nodes without field information). We can skip them.

**Comment by cshuaimin** [comment]
> A possible way is that a node is significant if a node is named node or is an unnamed node with field relative to its parent.

That's exactly what I'm doing here: https://github.com/cshuaimin/ssr.nvim/blob/main/lua/ssr/search.lua#L60-L70

**Comment by HerringtonDarkholme** [comment]
Thanks for the input @cshuaimin 
But I think it will fail the example of JavaScript getter?

**Comment by cshuaimin** [comment]
Yes it also fails on JavaScript getter. It's the same case as Rust underscore.

**Comment by HerringtonDarkholme** [comment]
It is fixed in 0.5.7!
